File: android/app/src/main/kotlin/com/TetherHealthLLC/tetherhealth/store/TetherStore.kt
package com.TetherHealthLLC.tetherhealth.store

import android.content.Context
import io.flutter.plugin.common.BinaryMessenger

/**
 * The Android half of the shell's key/value contract.
 *
 * Backed by a `SharedPreferences` file of its own, `tether_shell_state`, and
 * deliberately not by `unplug_shared_state`. That file is the Unplug module's:
 * the plugin, the watch service and the overlay all read it because they must
 * agree about the current tier and the override count. Nothing here is theirs —
 * this is the shell's enrolments, the person's answers and their lapse history.
 * Sharing one file would mean the overlay process could read which health
 * programmes somebody is enrolled in, for no benefit, and `purge()` on either
 * side would silently take the other's data with it.
 *
 * This class holds no product logic. It does not know what a snapshot contains
 * and must not learn: everything about shape, schema version and migration is
 * Dart, where it can be tested without a device.
 */
class TetherStore(
    context: Context,
    private val messenger: BinaryMessenger
) : TetherStoreApi {

    private val prefs =
        context.applicationContext.getSharedPreferences(NAME, Context.MODE_PRIVATE)

    init {
        TetherStoreApi.setUp(messenger, this)
    }

    /** Detaches the handler so a rebuilt engine does not talk to a dead one. */
    fun dispose() {
        TetherStoreApi.setUp(messenger, null)
    }

    override fun read(key: String): String? = prefs.getString(key, null)

    /**
     * Written with `commit()` rather than `apply()`.
     *
     * `apply()` returns immediately and writes on a background thread, which is
     * the right default and the wrong one here: the call that reaches this is
     * the app being backgrounded and flushing the last edit, which is precisely
     * the moment the process is most likely to be killed before an asynchronous
     * write lands. The channel call is already off the main thread as far as
     * Dart is concerned, so the cost is paid by nobody.
     */
    override fun write(key: String, value: String) {
        prefs.edit().putString(key, value).commit()
    }

    override fun remove(key: String) {
        prefs.edit().remove(key).commit()
    }

    /**
     * Wipes the whole file.
     *
     * Called by the `export_and_delete` hard delete, which promises the record
     * is gone rather than hidden. Total rather than selective, for the reason
     * `UnplugStore.purge()` gives about its own file: a partial wipe is the
     * kind of thing that looks compliant and is not. This file belongs to the
     * shell alone, so there is nothing here that a wipe should spare.
     */
    override fun clear() {
        prefs.edit().clear().commit()
    }

    companion object {
        /** Separate from `unplug_shared_state`. See the class comment. */
        private const val NAME = "tether_shell_state"
    }
}

File: ios/Runner/Store/TetherStore.swift
import Flutter
import Foundation

/// The iOS half of the shell's key/value contract.
///
/// Backed by `UserDefaults.standard` — the app's own container — and not by the
/// Unplug App Group. That is a deliberate boundary, not an oversight. The App
/// Group exists because `DeviceActivityMonitor`, `ShieldConfiguration` and
/// `ShieldAction` run in three separate processes and genuinely need to read
/// the module's tier and override counts. Nothing in this store is theirs: it
/// is the shell's enrolments, the person's answers and their lapse history.
/// Putting it in the group would hand three extensions readable access to a
/// record of which health programmes somebody is enrolled in, in exchange for
/// nothing at all, and the widest surface is the one that gets audited.
///
/// If an extension ever does need a value from here, the right move is to copy
/// that one value into the group, not to move this store into it.
///
/// The methods are synchronous because the contract is. `UserDefaults` is a
/// memory-backed cache after first touch and flushes on its own schedule, so
/// there is nothing here worth suspending a channel call for.
final class TetherStore: NSObject, TetherStoreApi {

    /// Everything this store writes is prefixed, so `clear()` can be honest
    /// about what it owns. `UserDefaults.standard` for the Runner also holds
    /// keys written by Flutter itself and by the system, and a wipe that took
    /// those with it would be a deletion feature that logs people out of
    /// unrelated things.
    private static let prefix = "tether.store."

    private let defaults: UserDefaults

    init(binaryMessenger: FlutterBinaryMessenger, defaults: UserDefaults = .standard) {
        self.defaults = defaults
        super.init()
        TetherStoreApiSetup.setUp(binaryMessenger: binaryMessenger, api: self)
    }

    func read(key: String) throws -> String? {
        defaults.string(forKey: Self.prefix + key)
    }

    func write(key: String, value: String) throws {
        defaults.set(value, forKey: Self.prefix + key)
    }

    func remove(key: String) throws {
        defaults.removeObject(forKey: Self.prefix + key)
    }

    /// Removes every key this store owns.
    ///
    /// Called by the `export_and_delete` hard delete, which promises the record
    /// is gone rather than hidden. It enumerates rather than removing one known
    /// key so that a second key added later is covered without anybody having
    /// to remember to extend the delete path — the failure mode of a selective
    /// wipe is that it keeps looking correct while it stops being complete.
    func clear() throws {
        for key in defaults.dictionaryRepresentation().keys where key.hasPrefix(Self.prefix) {
            defaults.removeObject(forKey: key)
        }
    }
}
